### PR TITLE
feat(process_tracker): make long standing payments failed

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] New feature

## Description

This feature will make the payments that have been stuck in `Pending` attempt status because of connection timed out or some other 5xx error. These payments will not have the connector transaction id. So after the last retry by process tracker, these will be moved to failed.

## How did you test it?

- Run producer and consumer locally.
- Simulate a connection timed out payment. Provide the connector url as https://google.com:81
- Wait for process tracker to make the status as `Failed` during the last retry.
Screenshot 2023-10-03 at 7 10 34 PM
Screenshot 2023-10-03 at 7 11 22 PM

- Check if a failed webhook is received.

Screenshot 2023-10-03 at 6 20 19 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

**Review by NishantJoshi00 — DISMISSED** [review]
If possible, it would be better to identify the payment whose status changed due to this last retry. to filter and identify them.
Other than that, looks good

**Comment by Narayanbhat166** [comment]
> If possible, it would be better to identify the payment whose status changed due to this last retry


Yes @NishantJoshi00. I have added a const error message for this purpose
`This Payment has been moved to failed as there is no response from the connector`

**Review by jarnura — APPROVED** [review]

**Review by NishantJoshi00 — DISMISSED** [review]

File: crates/router/src/workflows/payment_sync.rs [hunk 1]
```diff
@@ -93,15 +94,76 @@ impl ProcessTrackerWorkflow<AppState> for PaymentsSyncWorkflow {
...
+                            payment_data.payment_intent,
+                            payment_intent_update,
+                            merchant_account.storage_scheme,
+                        )
+                        .await
+                        .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+
+                    // Trigger the outgoing webhook to notify the merchant about failed payment
+                    let operation = operations::PaymentStatus;
+                    crate::utils::trigger_payments_webhook::<
```
Comment (jarnura) [inline crates/router/src/workflows/payment_sync.rs:151]:
```suggestion
                    utils::trigger_payments_webhook::<
```

File: crates/router/src/workflows/payment_sync.rs [hunk 2]
```diff
@@ -93,15 +94,76 @@ impl ProcessTrackerWorkflow<AppState> for PaymentsSyncWorkflow {
...
+                {
+                    let payment_intent_update = data_models::payments::payment_intent::PaymentIntentUpdate::PGStatusUpdate { status: api_models::enums::IntentStatus::Failed };
+                    let payment_attempt_update =
+                        data_models::payments::payment_attempt::PaymentAttemptUpdate::ErrorUpdate {
+                            connector: None,
+                            status: api_models::enums::AttemptStatus::AuthenticationFailed,
+                            error_code: None,
+                            error_message: None,
+                            error_reason: Some(Some(
+                                crate::consts::REQUEST_TIMEOUT_ERROR_MESSAGE_FROM_PSYNC.to_string(),
```
Comment (jarnura) [inline crates/router/src/workflows/payment_sync.rs:126]:
```suggestion
                                consts::REQUEST_TIMEOUT_ERROR_MESSAGE_FROM_PSYNC.to_string(),
```

File: crates/router/src/utils.rs [hunk 3]
```diff
@@ -669,3 +673,89 @@ pub fn add_apple_pay_payment_status_metrics(
...
+            .foreign_try_into()
+            .into_report()
+            .change_context(errors::ApiErrorResponse::WebhookProcessingFailure)
+            .attach_printable("payment event type mapping failed")?;
+
+        if let services::ApplicationResponse::JsonWithHeaders((payments_response_json, _)) =
+            payments_response
+        {
+            Box::pin(
+                crate::core::webhooks::create_event_and_trigger_appropriate_outgoing_webhook(
```
Comment (jarnura) [inline crates/router/src/utils.rs:745]:
```suggestion
                core::webhooks::create_event_and_trigger_appropriate_outgoing_webhook(
```

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Pull request merged**